example/lib/ide/problems_panel/problems_panel.dart

```diff
@@ -0,0 +1,78 @@
+import 'dart:io';
+
+import 'package:example/ide/problems_panel/diagnostics.dart';
+import 'package:example/lsp_exploration/lsp/lsp_client.dart';
+import 'package:flutter/material.dart';
+
+class ProblemsPanel extends StatefulWidget {
+  const ProblemsPanel({
+    super.key,
+    required this.lspClient,
+    this.sourceFile,
+  });
+
+  final LspClient lspClient;
+  final File? sourceFile;
+
+  @override
+  State<ProblemsPanel> createState() => _ProblemsPanelState();
+}
+
+class _ProblemsPanelState extends State<ProblemsPanel> {
+  List<Diagnostic>? diagnostics;
+  String? uri;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.lspClient.addNotificationListener(_onLspNotification);
+  }
+
+  void _onLspNotification(LspNotification notification) {
+    if (notification.method != 'textDocument/publishDiagnostics') {
+      return;
+    }
+    uri = notification.params['uri'] as String?;
+    final diagnosticsList = notification.params['diagnostics'] as List;
+    diagnostics = diagnosticsList.map((d) => Diagnostic.fromJson(d)).toList();
+    setState(() {});
```

Comment:
It won't be a functional difference, but to keep consistency with all Flutter code, set the state variables inside the. `setState` body.